src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -265,16 +197,101 @@ public void onMessage(String messageString, Session senderSession) {
     }
   }
 
+  /**
+   * Handle the first message from a workspace session, currently in purgatory,
+   * which should be a HELLO_WORKSPACE message.
+   *
+   * If it is a valid HELLO_WORKSPACE message, the workspace is authorized and added to the
+   * sessions map, and the new workspace count is broadcast to all workspaces (inclusive).
+   * Otherwise an error message is sent to the session and it is closed.
+   *
+   * The session is always removed from the purgatory set.
+   *
+   * @param session the session that sent the message.
+   * @param messageString the JSON string message.
+   */
+  private void handleHelloMessage(Session session, String messageString) {
+    // Regardless of if the message was good or bad, we remove the session from purgatory
+    // upon receipt of the first message.
+    purgatorySessions.remove(session);
+
+    // deserializeMessage() will close the connection if the message is invalid and
+    // not a HELLO_WORKSPACE message
+    var message = deserializeMessage(session, messageString, MessageType.WORKSPACE_HELLO);
+    if (message == null) {
+      Log.errorf(
+          "handleHelloMessage: Invalid message from purgatory session %s. Closed session.",
+          session.getId()
+      );
+    }
+
+    // Message is valid and is a HELLO_WORKSPACE message with a HelloBody.
+    var workspaceId = ((HelloBody) message.body()).workspaceId();
+
+    // Ensure that the header origin value corresponds to the
+    // workspace id in the body.
+    if (!validateOriginator(session, message, workspaceId)) {
+      sendErrorAndCloseSession(
+          session,
+          message.headers().id(),
+          "handleHelloMessage: Workspace id %s does not match originator value. Closing session.",
+          workspaceId
+      );
+      return;
+    }
+
+    // Cross-reference the workspace id against the known workspaces.
+
+    if (!knownWorkspacesBean.isKnownWorkspace(Long.valueOf(workspaceId))) {
+      sendErrorAndCloseSession(
+          session,
+          message.headers().id(),
+          "handleHelloMessage: Unknown workspace id %d. Closing session.",
+          workspaceId
+      );
+      return;
+    }
+
+    // Should be unique across all workspaces in session map.
+    for(var ws : sessions.values()) {
+      if (ws.processId() == workspaceId) {
+        sendErrorAndCloseSession(
+            session,
+            message.headers().id(),
+            "Workspace id %s already connected. Closing session.",
+            workspaceId
+        );
+        return;
+      }
+    }
+
+    // All good! Add the workspace to the sessions map and announce the count change.
+    sessions.put(session, new WorkspaceSession(workspaceId));
```

Comment:
What do you think about having the `WorkspaceSession` record hold a reference to the `Session` object? That way we can return a `WorkspaceSession` object and that also provides access to the `Session` object?

Reply:
Is doing that way now.